Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+                                                              blockingManager:self.blockingManager
+                                                                 dbConnection:self.editingDatabaseConnection
+                                                  hideUnreadMessagesIndicator:self.hasClearedUnreadMessagesIndicator
+                                                          lastUnreadIndicator:self.dynamicInteractions.unreadIndicator
+                                                               focusMessageId:self.focusMessageIdOnOpen
+                                                                 maxRangeSize:maxRangeSize];
+}
+
+- (nullable id<ConversationViewItem>)viewItemForUnreadMessagesIndicator
+{
+    for (id<ConversationViewItem> viewItem in self.viewItems) {
+        if (viewItem.unreadIndicator) {
+            return viewItem;
+        }
+    }
+    return nil;
+}
+
+- (void)clearUnreadMessagesIndicator
+{
+    OWSAssertIsOnMainThread();
+
+    // TODO: Remove by making unread indicator a view model concern.
+    id<ConversationViewItem> _Nullable oldIndicatorItem = [self viewItemForUnreadMessagesIndicator];
+    if (oldIndicatorItem) {
+        // TODO ideally this would be happening within the *same* transaction that caused the unreadMessageIndicator
+        // to be cleared.
+        [self.editingDatabaseConnection
+            asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+                [oldIndicatorItem.interaction touchWithTransaction:transaction];
+            }];
+    }
+
+    if (self.hasClearedUnreadMessagesIndicator) {
+        // ensureDynamicInteractionsForThread is somewhat expensive
+        // so we don't want to call it unnecessarily.
+        return;
+    }
+
+    // Once we've cleared the unread messages indicator,
+    // make sure we don't show it again.
+    self.hasClearedUnreadMessagesIndicator = YES;
+
+    if (self.dynamicInteractions.unreadIndicator) {
+        // If we've just cleared the "unread messages" indicator,
+        // update the dynamic interactions.
+        [self ensureDynamicInteractions];
+    }
+}
+
+#pragma mark - Storage access
+
+- (void)uiDatabaseDidUpdateExternally:(NSNotification *)notification
+{
+    OWSAssertIsOnMainThread();
+
+    OWSLogVerbose(@"");
+
+    if (self.delegate.shouldObserveDBModifications) {
+        // External database modifications can't be converted into incremental updates,
+        // so rebuild everything.  This is expensive and usually isn't necessary, but
+        // there's no alternative.
+        //
+        // We don't need to do this if we're not observing db modifications since we'll
+        // do it when we resume.
+        [self resetMappings];
+    }
+}
+
+- (void)uiDatabaseWillUpdate:(NSNotification *)notification
+{
+    [self.delegate conversationViewModelWillUpdate];
+}
+
+- (void)uiDatabaseDidUpdate:(NSNotification *)notification
```

Comment:
Like the old CVC, the CVM uses this method to respond to DB modifications.  Specifically, it:

* Updates its internal state (the view items).  This logic is almost identical.
* Fires update events to the CVC.  This is new.